client/extensions/woocommerce/app/products/product-form-variations-row.js

```diff
@@ -1,111 +1,231 @@
 /**
  * External dependencies
  */
-import React, { PropTypes } from 'react';
+import React, { Component, PropTypes } from 'react';
+import classNames from 'classnames';
+import Gridicon from 'gridicons';
 import { localize } from 'i18n-calypso';
+import { head } from 'lodash';
 
 /**
  * Internal dependencies
  */
 import formattedVariationName from 'woocommerce/lib/formatted-variation-name';
+import Button from 'components/button';
 import FormCurrencyInput from 'components/forms/form-currency-input';
 import FormDimensionsInput from 'woocommerce/components/form-dimensions-input';
 import FormTextInput from 'components/forms/form-text-input';
 import FormTextInputWithAffixes from 'components/forms/form-text-input-with-affixes';
+import ImagePreloader from 'components/image-preloader';
+import ProductImageUploader from 'woocommerce/components/product-image-uploader';
+import Spinner from 'components/spinner';
+
+class ProductFormVariationsRow extends Component {
+	static propTypes = {
+		product: PropTypes.object.isRequired,
+		variation: PropTypes.object.isRequired,
+		manageStock: PropTypes.bool,
+		onShowDialog: PropTypes.func,
+		editProductVariation: PropTypes.func.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		const { variation } = props;
+		const image = variation && variation.image || {};
+
+		this.state = {
+			id: image.id || null,
+			src: image.src || null,
+			placeholder: null,
+			transientId: null,
+			isUploading: false,
+		};
+	}
 
-const ProductFormVariationsRow = ( {
-	product,
-	variation,
-	editProductVariation,
-	onShowDialog,
-	translate,
-	manageStock,
-} ) => {
 	// TODO: Consildate the following set/toggle functions with a helper (along with the form-details functions).
-	const setPrice = ( e ) => {
+	setPrice = ( e ) => {
+		const { editProductVariation, product, variation } = this.props;
 		editProductVariation( product, variation, { regular_price: e.target.value } );
-	};
+	}
 
-	const setWeight = ( e ) => {
+	setWeight = ( e ) => {
+		const { editProductVariation, product, variation } = this.props;
 		editProductVariation( product, variation, { weight: e.target.value } );
-	};
+	}
 
-	const setDimension = ( e ) => {
+	setDimension = ( e ) => {
+		const { editProductVariation, product, variation } = this.props;
 		const dimensions = { ...variation.dimensions, [ e.target.name ]: e.target.value };
 		editProductVariation( product, variation, { dimensions } );
-	};
+	}
 
-	const setStockQuantity = ( e ) => {
+	setStockQuantity = ( e ) => {
+		const { editProductVariation, product, variation } = this.props;
 		const stock_quantity = Number( e.target.value ) >= 0 ? e.target.value : '';
 		editProductVariation( product, variation, { stock_quantity } );
-	};
+	}
 
-	const showDialog = () => {
+	showDialog = () => {
+		const { variation, onShowDialog } = this.props;
 		onShowDialog( variation.id );
-	};
+	}
+
+	onSelect = ( files ) => {
+		const file = head( files );
+		this.setState( {
```

Comment:
`setState` is a [React method](https://facebook.github.io/react/docs/react-component.html#setstate) for changing internal component state (changing the state previously set in `this.state`).